test/verify/check-machines

```diff
@@ -3435,6 +3436,74 @@ class TestMachines(MachineCase, StorageHelpers, NetworkHelpers):
         b.click('#deactivate-network-test_network4-{0}'.format(connectionName)) # Deactivate transient network
         self.waitNetworkRow("test_network4", connectionName, False) # Check it's not present after deactivation
 
+    def testSnapshots(self):
+        # Checks if difference of @time1 and @time2 is not greater than @difference (in seconds)
+        def checkTimeDiff(time1, time2, difference):
+            tmp = time2.split(' ') # split "Today 13:13:13" into day and time
+            if not tmp[1].startswith('00:') and not tmp[0] == "Today":
+                return False
+
+            diff = datetime.strptime(time1, '%H:%M:%S') - datetime.strptime(tmp[1] + tmp[2], '%I:%M%p')
+            return -difference < diff.total_seconds() < difference
+
+        b = self.browser
+        m = self.machine
+
+        self.startVm("subVmTest1")
+
+        snapshots_present = m.execute("gdbus introspect --xml --system --dest org.libvirt --object-path /org/libvirt/QEMU | grep snapshot || true")
+        if not snapshots_present.strip(): # libvirt-dbus doesn't have snapshot APIs
+            return
+
+        self.login_and_go("/machines")
+        b.wait_in_text("body", "Virtual Machines")
+        self.waitVmRow("subVmTest1")
+        self.toggleVmRow("subVmTest1")
+        b.wait_in_text("#vm-subVmTest1-state", "running")
+        b.click("#vm-subVmTest1-snapshots") # open the "Disks" subtab
+        b.wait_in_text(".no-snapshots", "No snapshots")
+
+        # Check snapshot for running VM
+        m.execute("virsh detach-disk --domain subVmTest1 --target vda --persistent") # vda is raw disk, which are not supported by internal snapshots
+        m.execute("qemu-img create -f qcow2 /var/lib/libvirt/images/foobar.qcow2 1M")
+        m.execute("virsh attach-disk --domain subVmTest1 --source /var/lib/libvirt/images/foobar.qcow2 --target vdb --persistent")
+        time1 = datetime.now().strftime("%H:%M:%S")
+        m.execute("virsh snapshot-create-as --domain subVmTest1 --name snapshot1 --description 'Description of snapshot1' --disk-only")
+
+        b.reload()
+        b.enter_page('/machines')
+        b.wait_in_text("body", "Virtual Machines")
+        self.toggleVmRow("subVmTest1")
+        b.click("#vm-subVmTest1-snapshots")
+
+        b.wait_in_text("#vm-subVmTest1-snapshot-0-name", "snapshot1")
+        b.wait_in_text("#vm-subVmTest1-snapshot-0-description", "Description of snapshot1")
+        b.wait_in_text("#vm-subVmTest1-snapshot-0-type", "disk-snapshot")
+        b.wait_in_text("#vm-subVmTest1-snapshot-0-parent", "No parent")
+        time2 = b.text("#vm-subVmTest1-snapshot-0-date")
+        self.assertTrue(checkTimeDiff(time1, time2, 60))
+
+        # Check snapshot for shutoff VM
+        b.click("#vm-subVmTest1-action-kebab button")
+        b.click("#vm-subVmTest1-forceOff")
+        b.wait_in_text("#vm-subVmTest1-state", "shut off")
```

Comment:
"shut off" is the state of a Machine. It doesn't make sense as a snapshot state -- a snapshot is either disk-only or includes runtime state (RAM).

Reply:
I think running/shutoff comes from virt-manager.
RHV, WMWare and OpenShift all threats this as snapshotting the memory.
* VMWare https://vmexchange.files.wordpress.com/2013/01/vmware-snapshot-window.png
* RHV https://www.tecmint.com/wp-content/uploads/2016/01/Snapshot-In-Preview-State.png
* OpenShift https://raw.githubusercontent.com/Ranelim/openshift-origin-design/60f28713cc45f6e42a052c2ae442fb2b2674f2e2/web-console/knikubevirt/snapshots/img/3-02.jpg

@martinpitt did we ever come to a good conclusion regarding this in the thread with Matthew Carleton and others late last year?

Reply:
@andreasn : None of the screenshots have "shut off" though, so they are not evindence that this is being used somehwere else. Indeed we haven't heard back from Matthew about that topic. Indeed we should try to be consistent with virt-manager and other products, but IMO "shut off" is a very confusing state for a snapshot..

Reply:
Since there is no consensus, I guess let's keep it for now. It's true that it doesn't refer to the state of snapshot itself, rather state of virtual machine at the moment of snapshot capture (But I believe that's rather important information too to show).

Reply:
I agree that shut off is quite odd in this context. It's not running, certainly, but snapshots should never be running.

Brainstorming possible labels:

- bootable / resumable
- disk-only / disk + memory

Or omit the label on the disk-only snapshots and add a label when they have RAM snapshots too:
- includes state
- disk + memory
- resumable

Reply:
Perhaps it's better to split this semantic into two columns:
1. **Snapshot type** - possible values are: disk only snapshot, full system snapshot (disk+memory)
2. **VM's state** at the moment of capture - possible values are: running, blocked, paused, shutdown, shut off, crashed

Reply:
> I agree that shut off is quite odd in this context. It's not running, certainly, but snapshots should never be running.
> 
> Brainstorming possible labels:
> 
> * bootable / resumable
> * disk-only / disk + memory
> 
> Or omit the label on the disk-only snapshots and add a label when they have RAM snapshots too:
> 
> * includes state
> * disk + memory
> * resumable

I agree with having a label of only disk-only / disk + memory. When it's disk-only, ommit VM state, when it's disk + memory, could show also VM state.
And don't really understand what do you mean by resumable,

Reply:
So did we come to any conclusion on this issue?

Reply:
Possible values for domain snapshot state are:
running
blocked
paused
shutdown
shutoff
crashed
disk-snapshot

The first 6 refer to the state of VM when was snapshot taken. Last one is when user specifies they only want to capture the disk content, so snapshot doesn't remember what is the state of VM.
Perhaps it's redundant to add another column just for this one snapshot state, since most of the time users will use snapshots which remember VM's state. So perhaps the solution could be:

Rename "Type" column to "VM state"
Show the above mentioned values. In case if it's "disk-snapshot", the string shown of "VM state" column could be something like "No VM state tracked (disk only snapshot)". @garrett @andreasn what do you think?